### PR TITLE
CSK cluster support RBAC

### PR DESCRIPTION
**Pull request by anserhyb**

### CODE REVIEW

File: alicloud/resource_alicloud_cs_kubernetes_permissions.go [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+		},
+
+		Schema: map[string]*schema.Schema{
+			"uid": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"permissions": {
+				Optional: true,
+				Type:     schema.TypeList,
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_cs_kubernetes_permissions.go:42]: I think its type should be TypeSet

File: alicloud/resource_alicloud_cs_kubernetes_permissions.go [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+	if err != nil {
+		return WrapErrorf(err, DefaultErrorMsg, ResourceName, "InitializeClient", err)
+	}
+
+	// Query existing permissions, DescribeUserPermission
+	uid := d.Get("uid").(string)
+	_, _ = describeUserPermission(client, uid)
+	//if _err != nil {
+	//	return WrapErrorf(err, DefaultErrorMsg, ResourceName, "DescribeUserPermission", err)
+	//}
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_cs_kubernetes_permissions.go:111]: If there is an error, it should be return.

File: alicloud/resource_alicloud_cs_kubernetes_permissions.go [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+}
+
+func resourceAlicloudCSKubernetesPermissionsRead(d *schema.ResourceData, meta interface{}) error {
+	client, err := initCsClient(meta)
+	if err != nil {
+		return WrapErrorf(err, DefaultErrorMsg, ResourceName, "InitializeClient", err)
+	}
+
+	// Query existing permissions, DescribeUserPermission
+	uid := d.Get("uid").(string)
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_cs_kubernetes_permissions.go:107]: uid can be got by `d.Id()`

File: alicloud/resource_alicloud_cs_kubernetes_permissions.go [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+
+func resourceAlicloudCSKubernetesPermissionsUpdate(d *schema.ResourceData, meta interface{}) error {
+	d.Partial(true)
+
+	client, err := initCsClient(meta)
+	if err != nil {
+		return WrapErrorf(err, DefaultErrorMsg, ResourceName, "InitializeClient", err)
+	}
+
+	uid := d.Get("uid").(string)
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_cs_kubernetes_permissions.go:125]: still applies

File: alicloud/resource_alicloud_cs_kubernetes_permissions.go [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+
+	// Remove up all permissions owned by the user
+	err = grantPermissions(client, uid, []*cs.GrantPermissionsRequestBody{})
+	if err != nil {
+		return WrapErrorf(err, DefaultErrorMsg, ResourceName, "RemoveUserPermissions", err)
+	}
+	return nil
+}
+
+func initCsClient(meta interface{}) (*cs.Client, error) {
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_cs_kubernetes_permissions.go:163]: I think it should be inited in the client.go

File: website/docs/r/cs_kubenetes_permissions.html.markdown [hunk 6]
```diff
@@ -0,0 +1,448 @@
...
+resource "alicloud_vpc" "default" {
+  vpc_name                     = "${var.name}"
+  cidr_block                   = "10.1.0.0/21"
+}
+
+resource "alicloud_vswitch" "default" {
+  vswitch_name                 = "${var.name}"
+  vpc_id                       = "${alicloud_vpc.default.id}"
+  cidr_block                   = "10.1.1.0/24"
+  availability_zone            = "${data.alicloud_zones.default.zones.0.id}"
```
Comment (xiaozhu36) [inline website/docs/r/cs_kubenetes_permissions.html.markdown:51]: Please use zone_id

File: website/docs/r/cs_kubenetes_permissions.html.markdown [hunk 7]
```diff
@@ -0,0 +1,448 @@
...
+variable "name" {
+  default = "custom-name"
+}
+
+data "alicloud_zones" default {
+  available_resource_creation  = "VSwitch"
+}
+
+data "alicloud_instance_types" "default" {
+  availability_zone          = "${data.alicloud_zones.default.zones.0.id}"
```
Comment (xiaozhu36) [inline website/docs/r/cs_kubenetes_permissions.html.markdown:36]: There is no need to using grammar `"${}"` and just using `data.alicloud_zones.default.zones.0.id` or `var.name` is ok.

File: website/docs/r/cs_kubenetes_permissions.html.markdown [hunk 8]
```diff
@@ -0,0 +1,448 @@
...
+
+This resource will help you implement RBAC authorization for the kubernetes cluster. 
+
+-> **NOTE:** Please make sure that the target RAM user has been granted a RAM policy with at least read-only permission of the target cluster in the RAM console. Otherwise, the `ErrorRamPolicyConfig` error will be returned. 
+For more information about how to authorize a RAM user by attaching RAM policies, see [Create a custom RAM policy](https://www.alibabacloud.com/help/doc-detail/86485.htm).
+
+-> **NOTE:** If you call this operation as a RAM user, make sure that this RAM user has the permissions to grant other RAM users the permissions to manage ACK clusters. Otherwise, the `StatusForbidden` or `ForbiddenGrantPermissions` errors will be returned. For more information, see [Use a RAM user to grant RBAC permissions to other RAM users](https://www.alibabacloud.com/help/faq-detail/119035.htm).
+
+-> **NOTE:** This operation overwrites the permissions that have been granted to the specified RAM user. When you call this operation, make sure that the required permissions are included.
+
```
Comment (xiaozhu36) [inline website/docs/r/cs_kubenetes_permissions.html.markdown:20]: Please adds an availability note.

**Review by xiaozhu36 — CHANGES_REQUESTED** [review]

File: alicloud/data_source_alicloud_cs_kubernetes_permissions.go [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func dataAlicloudCSKubernetesPermissionsRead(d *schema.ResourceData, meta interface{}) error {
+	client, err := cs.NewClient(&openapi.Config{
```
Comment (xiaozhu36) [inline alicloud/data_source_alicloud_cs_kubernetes_permissions.go:62]: Please move the client to connectivity.

File: alicloud/resource_alicloud_cs_kubernetes_permissions.go [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceAlicloudCSKubernetesPermissionsCreate(d *schema.ResourceData, meta interface{}) error {
+	client, err := cs.NewClient(&openapi.Config{
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_cs_kubernetes_permissions.go:78]: still applies

**Review by xiaozhu36 — CHANGES_REQUESTED** [review]

File: alicloud/resource_alicloud_cs_kubernetes_permissions.go (hunk 2 above)
Comment (anserhyb) [reply alicloud/resource_alicloud_cs_kubernetes_permissions.go]: 这段去掉了

File: alicloud/resource_alicloud_cs_kubernetes_permissions.go (hunk 1 above)
Comment (anserhyb) [reply alicloud/resource_alicloud_cs_kubernetes_permissions.go]: 将类型改为TypeSet了

File: alicloud/resource_alicloud_cs_kubernetes_permissions.go (hunk 3 above)
Comment (anserhyb) [reply alicloud/resource_alicloud_cs_kubernetes_permissions.go]: 替换为了通过 `d.Id()`来获取了。

File: alicloud/resource_alicloud_cs_kubernetes_permissions.go (hunk 4 above)
Comment (anserhyb) [reply alicloud/resource_alicloud_cs_kubernetes_permissions.go:111]: 替换为了通过 d.Id()来获取了。

File: alicloud/resource_alicloud_cs_kubernetes_permissions.go (hunk 5 above)
Comment (anserhyb) [reply alicloud/resource_alicloud_cs_kubernetes_permissions.go]: 将这段逻辑支持，在connectivity里支持了。

File: website/docs/r/cs_kubenetes_permissions.html.markdown (hunk 6 above)
Comment (anserhyb) [reply website/docs/r/cs_kubenetes_permissions.html.markdown]: 已经修改为最新用法

File: website/docs/r/cs_kubenetes_permissions.html.markdown (hunk 7 above)
Comment (anserhyb) [reply website/docs/r/cs_kubenetes_permissions.html.markdown]: 已经更正：https://github.com/aliyun/terraform-provider-alicloud/pull/3495/commits/8883421d5b72afc453a56ebaad5167bce892e0c2

File: website/docs/r/cs_kubenetes_permissions.html.markdown (hunk 8 above)
Comment (anserhyb) [reply website/docs/r/cs_kubenetes_permissions.html.markdown]: https://github.com/aliyun/terraform-provider-alicloud/pull/3495/commits/8883421d5b72afc453a56ebaad5167bce892e0c2

File: alicloud/data_source_alicloud_cs_kubernetes_permissions.go (hunk 9 above)
Comment (anserhyb) [reply alicloud/data_source_alicloud_cs_kubernetes_permissions.go]: update in :https://github.com/aliyun/terraform-provider-alicloud/pull/3495/commits/cf231d5079bb6e37dae54dbec0743935e6fee91c

File: alicloud/resource_alicloud_cs_kubernetes_permissions.go (hunk 10 above)
Comment (anserhyb) [reply alicloud/resource_alicloud_cs_kubernetes_permissions.go]: update in: https://github.com/aliyun/terraform-provider-alicloud/pull/3495/commits/cf231d5079bb6e37dae54dbec0743935e6fee91c

File: alicloud/resource_alicloud_cs_kubernetes_permissions.go [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+	// Grant Permissions
+	// If other permissions with this right already exist, the existing permissions will be merged
+	grantPermissionsRequest := buildPermissionArgs(d)
+	err = grantPermissionsForAddPerm(client, uid, grantPermissionsRequest)
+	if err != nil {
+		return WrapErrorf(err, DefaultErrorMsg, ResourceName, "GrantPermissions", AliyunTablestoreGoSdk)
+	}
+
+	addDebug("GrantPermissions", grantPermissionsRequest, err)
+	_ = d.Set("uid", uid)
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_cs_kubernetes_permissions.go:93]: The line 93 is useless and please delete it.

File: alicloud/resource_alicloud_cs_kubernetes_permissions.go [hunk 12]
```diff
@@ -0,0 +1,382 @@
...
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(90 * time.Minute),
+			Update: schema.DefaultTimeout(60 * time.Minute),
+			Delete: schema.DefaultTimeout(60 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"uid": {
+				Type:     schema.TypeString,
+				Required: true,
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_cs_kubernetes_permissions.go:36]: the uid should be ForceNew

**Review by xiaozhu36 — CHANGES_REQUESTED** [review]

File: alicloud/resource_alicloud_cs_kubernetes_permissions.go (hunk 12 above)
Comment (anserhyb) [reply alicloud/resource_alicloud_cs_kubernetes_permissions.go:36]: https://github.com/aliyun/terraform-provider-alicloud/pull/3495/commits/a4f687532155d16a0c16e7a86406dd17ecc353aa

File: alicloud/resource_alicloud_cs_kubernetes_permissions.go (hunk 11 above)
Comment (anserhyb) [reply alicloud/resource_alicloud_cs_kubernetes_permissions.go]: https://github.com/aliyun/terraform-provider-alicloud/pull/3495/commits/a4f687532155d16a0c16e7a86406dd17ecc353aa

**Pull request merged**